nbformat/notebooknode.py

```diff
@@ -1,23 +1,50 @@
 """NotebookNode - adding attribute access to dicts"""
 
 from ipython_genutils.ipstruct import Struct
+from collections import Mapping
+
 
 class NotebookNode(Struct):
     """A dict-like node with attribute-access"""
-    pass
+
+    def __setitem__(self, key, value):
+        if isinstance(value, Mapping):
+            value = from_dict(value)
```

Comment:
The reasoning for including it in `from_dict` is that that should be a safety check extant everywhere `from_dict`  is used. 

I'm guessing we won't want to change `from_dict` as used in v1/v2/v3 but I can also propagate the change there.